### PR TITLE
Allowing config option to set which headers to extract

### LINKED ISSUE #984: Add Ability to specify how deep page headers should go
## Feature request
It would be nice to have the ability to specify how deep the `extractHeaders` method goes.
Currently, it has a hard coded value of `['h2', 'h3']`
Our current markdown files use `h4` as well in order to handle some table of contents for the page in sidebar form.

#### What problem does this feature solve?
When doing a table of contents in the sidebar, this feature would allow you to specify how deep the TOC should go based on headers.

#### What does the proposed API look like?
config.js
```
module.exports = {
    themeConfig: {
        parseHeaders: ['h2','h3','h4']
    }
}
```

#### How should this be implemented in your opinion?
The above would be a themeConfig option and will replace the default hard coded value located at 
https://github.com/vuejs/vuepress/blob/master/packages/%40vuepress/core/lib/prepare/Page.js#L92

#### Are you willing to work on this yourself?**
Yes

### PR DESCRIPTION
**Pull request by bretterer**
**Summary**
This PR resolves #984 
This PR replaces #1004 

Currently the header parsing is hard coded to `['h2', 'h3']`. This code changes that to give the ability to configure this in the `themeOptions` config.

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [x] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI of default theme, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [ ] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

**Comment by bretterer** [comment]
This PR is a cleaned up version of the pr at #1004  

There was an issue with the rebase I did causing a bunch of extra files being merged in that were not part of the actual PR. Tried to clean them up, but was not able to.

@ulivz

**Comment by bretterer** [comment]
Happy New Year @ulivz 

Just circling back around to this PR (and #1004) to see if this is ready to merge in now.  Really need to have this to finish up a project

File: packages/@vuepress/markdown-loader/index.js [hunk 1]
```diff
@@ -42,7 +43,11 @@ module.exports = function (src) {
 
   if (!isProd && !isServer) {
     const inferredTitle = inferTitle(frontmatter.data, frontmatter.content)
-    const headers = extractHeaders(content, ['h2', 'h3'], markdown)
+    let headersToExtract = ['h2', 'h3']
+    if (siteConfig.markdown && siteConfig.markdown.extractHeaders) {
+      headersToExtract = siteConfig.markdown.extractHeaders
+    }
+    const headers = extractHeaders(content, headersToExtract, markdown)
```
Comment (ulivz) [inline packages/@vuepress/markdown-loader/index.js]: I remember I have said that `siteConfig` shouldn't be passed to ·`markdown-loader`

Comment (bretterer) [reply packages/@vuepress/markdown-loader/index.js]: Should I just change this to pass in `siteConfig.markdown` instead of all of `siteConfig`?

**Comment by f3ltron** [comment]
We should take care about the search too with that to be sure all is working well

**Comment by f3ltron** [comment]
Hello i'm closing this PR as it's now replace by this one #1730 Thank's for your work all !

**Pull request closed without merging**